lib/Dialect/QUIR/Transforms/ExtractCircuits.cpp

```diff
@@ -0,0 +1,333 @@
+                                mlir::quir::getPhysicalIdAttrName()),
+                physicalId)}));
+      }
+    } else {
+      argumentIndex = search->second;
+    }
+
+    mapper.map(operand, currentCircuitOp.getArgument(argumentIndex));
+  }
+  auto *newOp = circuitBuilder.clone(*currentOp, mapper);
+
+  outputTypes.append(newOp->getResultTypes().begin(),
+                     newOp->getResultTypes().end());
+  outputValues.append(newOp->getResults().begin(), newOp->getResults().end());
+  originalResults.append(currentOp->getResults().begin(),
+                         currentOp->getResults().end());
+
+  eraseList.push_back(currentOp);
+}
+
+void ExtractCircuitsPass::endCircuit(
+    Operation *firstOp, Operation *lastOp, OpBuilder topLevelBuilder,
+    OpBuilder circuitBuilder, llvm::SmallVector<Operation *> &eraseList) {
+
+  LLVM_DEBUG(llvm::dbgs() << "Ending circuit " << currentCircuitOp.getSymName()
+                          << "\n");
+
+  circuitBuilder.create<mlir::quir::ReturnOp>(lastOp->getLoc(), outputValues);
+
+  // change the input / output types for the quir.circuit
+  auto opType = currentCircuitOp.getFunctionType();
+  currentCircuitOp.setType(topLevelBuilder.getFunctionType(
+      /*inputs=*/opType.getInputs(),
+      /*results=*/ArrayRef<Type>(outputTypes)));
+
+  std::sort(phyiscalIds.begin(), phyiscalIds.end());
+  currentCircuitOp->setAttr(
+      mlir::quir::getPhysicalIdsAttrName(),
+      topLevelBuilder.getI32ArrayAttr(ArrayRef<int>(phyiscalIds)));
+
+  // insert call_circuit
+  OpBuilder builder(firstOp);
+  newCallCircuitOp = builder.create<mlir::quir::CallCircuitOp>(
+      currentCircuitOp->getLoc(), currentCircuitOp.getSymName(),
+      TypeRange(outputTypes), ValueRange(inputValues));
+
+  // remap uses
+  assert(originalResults.size() == newCallCircuitOp->getNumResults() &&
+         "number of results does not match");
+  for (uint cnt = 0; cnt < newCallCircuitOp->getNumResults(); cnt++) {
+    originalResults[cnt].replaceAllUsesWith(newCallCircuitOp->getResult(cnt));
+    assert(originalResults[cnt].use_empty() && "usage expected to be empty");
+  }
+
+  // erase operations
+  while (!eraseList.empty()) {
+    auto *op = eraseList.back();
+    eraseList.pop_back();
+    assert(op->use_empty() && "operation usage expected to be empty");
+    LLVM_DEBUG(llvm::dbgs() << "Erasing: ");
+    LLVM_DEBUG(op->dump());
+    op->erase();
+  }
+}
+
+void ExtractCircuitsPass::processOps(Operation *currentOp,
+                                     OpBuilder topLevelBuilder,
+                                     OpBuilder circuitBuilder) {
+
+  llvm::SmallVector<Operation *> eraseList;
+
+  Operation *firstQuantumOp = nullptr;
+
+  // Handle Shot Loop delay differently
+  if (isa<quir::DelayOp>(currentOp) &&
+      isa<qcs::ShotInitOp>(currentOp->getNextNode())) {
+    // skip pass shot init
+    currentOp = currentOp->getNextNode()->getNextNode();
+  }
+
+  while (currentOp) {
```

Comment:
I think a few comments in the body of this loop would help in understanding the flow.

Reply:
Added comments with 41febe7. 
Also remove special handling of `qcs.DelayCycleOp` and verified test case binary output vs without circuits.